# 01. Importing libraries

In [56]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 02. Path

In [57]:
#Creating a path varaible - Shortcut for Importing Data Files
path = r'C:\Users\Mallika\Documents\Career Foundry\Module2- Data Immersion\Achievement 4\June2025_Instacart Basket Analysis'

# 03. Importing Data


In [58]:
#Import the products Dataset
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'products.csv'), index_col = False)
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [59]:
#Import the orders Dataset
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)
df_ords

Unnamed: 0  order_id  user_id  order_number  order_dow  \
0                 0   2539329        1             1          2   
1                 1   2398795        1             2          3   
2                 2    473747        1             3          3   
3                 3   2254736        1             4          4   
4                 4    431534        1             5          4   
...             ...       ...      ...           ...        ...   
3421078     3421078   2266710   206209            10          5   
3421079     3421079   1854736   206209            11          4   
3421080     3421080    626363   206209            12          1   
3421081     3421081   2977660   206209            13          1   
3421082     3421082    272231   206209            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

# 04. Data Consistency Checks

01. Mixed-Type Data


In [60]:
#Creating a mix data type column in a test dataframe
df_test = pd.DataFrame()

In [61]:
#The mixed column if filled with different types of data 
df_test['mix'] = ['a', 'b', 1, True]

In [62]:
#Creating fuction to search for a mixed data type in a column
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\Mallika\AppData\Local\Temp\ipykernel_2888\2911751512.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [63]:
#Converting the mixed column data type to String as most of the data in the mixed column is of a string.
df_test['mix'] = df_test['mix'].astype('str')

In [64]:
df_test['mix'].dtype


dtype('O')

02. Missing Values


In [65]:
#Finding missing values in the products dataset
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [66]:
#Creating a datafarme to subset the data in the df_prods where there  missing values are found  in the product_name column
df_nan = df_prods[df_prods['product_name'].isnull() == True ]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [67]:
#Running explanatory analysis on the df_products dataframe
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [68]:
#Fidning the median of the df_prods dataframe
df_prods['prices'].median() 

7.1

In [69]:
#Finding outliers in the df_prods dataframe.
df_prods[df_prods['prices'] > df_prods['prices'].mean() + (2 * df_prods['prices'].std())]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

In [70]:
#Replace the outliers values in the prices column with the meadian value of 7.1.
df_prods = df_prods.replace({"prices":{14900.0:7.1, 99999.0:7.1}})

In [71]:
#Checking the df_prods dataframe after imputation
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      7.682245
std    14343.717401     38.316774       5.850282      4.200159
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000     25.000000

## Observation: 
After conducting exploratory analysis on the df_products DataFrame, outliers were detected in the prices column. The column had a mean of 9.99 and a high standard deviation of 453.6, which is well beyond two standard deviations from the mean—indicating the presence of extreme values.

Specifically, two outlier values were identified: 14,900 and 99,999. These values were more than two standard deviations above the mean. To address this, they were replaced with the median price of 7.1.

Given the wide range between the minimum value 1 and the maximum value 99999, using the median is more appropriate than the mean for handling outliers, as it is less influenced by extreme values.

After replacing the outliers with the median, the overall mean of the prices column dropped to 7.6, reflecting a more accurate representation of the dataset.

In [72]:
#Checking the number of rows & columns in df_prods dataframe
df_prods.shape

(49693, 5)

In [73]:
#Creating a new dataframe as a subset of df_prods dataframe where the missing values in the product_name variable are excluded
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

03. Duplicate Values


In [74]:
#Fidning duplicates in the df_prods_clean dataframe and subset those duplicate data and store them in the df_dups -new dataframe 
df_dups = df_prods_clean[df_prods_clean.duplicated()]
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [75]:
#Checking the number of rows in the df_prods_clean datafrane
df_prods_clean.shape

(49677, 5)

In [76]:
#Creating a new datafrme where there is no duplicate values
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [77]:
df_prods_clean_no_dups.shape

(49672, 5)

## 05: Task 

In [78]:
#Step2: Run the df.describe() function on your df_ords dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

Observation: An assessment was conducted to verify the accuracy of the columns in the df_ords DataFrame. Upon reviewing the values in each column, no anomalies or unreasonable entries were identified. All values appear to fall within acceptable and interpretable ranges for their respective variables.

In [79]:
#Step 3: mixed-type data in your df_ords dataframe
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  order_dow  order_hour_of_day  \
0           0   2539329        1             1          2                  8   
1           1   2398795        1             2          3                  7   
2           2    473747        1             3          3                 12   
3           3   2254736        1             4          4                  7   
4           4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

Observation: There is no mixed-type data found in any of the columns/variables in the df_ords dataframe.

In [80]:
# Step4 Running the fucntion to search for a mixed data type in a column in the df_ords dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)      

C:\Users\Mallika\AppData\Local\Temp\ipykernel_2888\3419648606.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Mallika\AppData\Local\Temp\ipykernel_2888\3419648606.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Mallika\AppData\Local\Temp\ipykernel_2888\3419648606.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Mallika\AppData\Local\Temp\ipykernel_2888\3419648606.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Mallika\AppData

In [81]:
#Checking the data types of all the columns in the df_ords dataframe
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


Observation: No mixed-type data was found in any of the columns in the df_ords DataFrame. This was confirmed by using a function to detect mixed types, which returned NULL, and further validated using the info() function. All columns contain consistent data types, with no evidence of mixed-type values across any variables.

In [82]:
#Step5: Run a check for missing values in your df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

df_ords

In [83]:
#Find out the the missing values based on user_id count summarized 
df_ords[df_ords['days_since_prior_order'].isnull()].groupby('user_id').count().describe()

Unnamed: 0  order_id  order_number  order_dow  order_hour_of_day  \
count    206209.0  206209.0      206209.0   206209.0           206209.0   
mean          1.0       1.0           1.0        1.0                1.0   
std           0.0       0.0           0.0        0.0                0.0   
min           1.0       1.0           1.0        1.0                1.0   
25%           1.0       1.0           1.0        1.0                1.0   
50%           1.0       1.0           1.0        1.0                1.0   
75%           1.0       1.0           1.0        1.0                1.0   
max           1.0       1.0           1.0        1.0                1.0   

       days_since_prior_order  
count                206209.0  
mean                      0.0  
std                       0.0  
min                       0.0  
25%                       0.0  
50%                       0.0  
75%                       0.0  
max                       0.0

Observation: A total of 206,209 missing values were identified in the days_since_prior_order column of the original df_ords DataFrame. 
Given that the DataFrame contains 3,421,083 rows and 7 columns, and that there are exactly 206,209 unique user_id values, this indicates that each user has one missing entry in this column.
This pattern suggests that the missing values correspond to the first order placed by each user, where no prior order exists to calculate the number of days since. Therefore, these missing values are expected and contextually appropriate.
As a result, it is unnecessary to remove them from the dataset.

Step6:
Observation: There is no need to remove these missing values, since each unique user_id should have only one missing value in the days_since_prior_order column, and that missing value belongs to the very first order of each unique user_id.

In [84]:
#Finding duplicates in the df_ords dataframe
df_ords[df_ords.duplicated()]

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

Observation for Step 7 & 8,
There are no duplicate values found in the df_ords dataframe.

In [85]:
#Chekcing the df_ords dataframe
df_ords

Unnamed: 0  order_id  user_id  order_number  order_dow  \
0                 0   2539329        1             1          2   
1                 1   2398795        1             2          3   
2                 2    473747        1             3          3   
3                 3   2254736        1             4          4   
4                 4    431534        1             5          4   
...             ...       ...      ...           ...        ...   
3421078     3421078   2266710   206209            10          5   
3421079     3421079   1854736   206209            11          4   
3421080     3421080    626363   206209            12          1   
3421081     3421081   2977660   206209            13          1   
3421082     3421082    272231   206209            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [86]:
#Dropping the column "Unnamed: 0" from the df_ords dataframe and assinged it to a new dataframe "df_ords_clean"
df_ords_clean = df_ords.drop('Unnamed: 0', axis=1)
df_ords_clean

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

Observation: After checking the df_ords dataframe, it was noticed that there is an unnecessary column called "Unnamed: 0" that needs to be dropped.


In [87]:
#Export the cleaned datafrme for df_prods
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [88]:
#Export the cleaned datafrme for df_ords
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

## Export


df_prods: final clean version = df_prods_clean_no_dups, and saved as products_checked.csv

df_ords: final clean version = df_ords_clean, and saved as orders_checked.csv